### PR TITLE
post: xnet protocol

### PR DESCRIPTION
**Pull request by roman-kashitsyn**

### CODE REVIEW

File: posts/08-ic-xnet.html.pm [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  }
+  ◊p{◊(embed-svg "images/08-xnet-endpoint.svg")}
+}
+}
+
+◊section{
+◊section-title["message-streams"]{Message streams}
+◊p{
+  Messages start their journey in the output queues of canisters hosted on the source subnet.
+  The subnet sorts these messages based on the destination subnet and conflates them into a flat message stream (one stream per destination subnet).
```
Comment (alin-at-dfinity) [inline posts/08-ic-xnet.html.pm:84]: "Conflate" has a negative connotation in my head. Wikipedia seems to agree. I'd go with "merges" or "routes".

Also "a flat message stream" -> "flat message streams".

File: posts/08-ic-xnet.html.pm [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    The full range of message indices in the forward stream ◊math{X → Y}.
+  }
+  ◊li{
+    The signals for the ◊em{reverse} stream (◊math{Y → X}): for each message index in the reverse stream, ◊math{Y} tells whether ◊math{X} can garbage collect the message (an ◊code{ACK} signal) or should reschedule the message (a ◊code{REJECT} signal).
+    A ◊code{REJECT} signal indicates that the destination canister moved, so ◊math{X} should insert the message into another stream.
+  }
+}
+◊p{
+  Signals solve the issue of collecting obsolete messages, but they introduce another problem now we also need to garbage-collect signals!
+  Luckily, we already have all the information we need: we keep signals only for messages that are still present in the remote stream.
```
Comment (alin-at-dfinity) [inline posts/08-ic-xnet.html.pm:144]:
```suggestion
  Luckily, we already have all the information we need: we keep signals only for messages that are still present in the reverse stream.
```

File: posts/08-ic-xnet.html.pm [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  ◊marginnote["mn-signals"]{
+    Two subnets, ◊math{X} and ◊math{Y}, communicating over the XNet protocol.
+    In previous communications, subnet ◊math{Y} consumed messages ◊math{[0,10)} from ◊math{X}'s stream and has recently received messages ◊math{10} and ◊math{11}.
+    Now, subnet ◊math{X} receives a prefix of ◊math{Y}'s stream and the matching header and removes messages ◊math{10} and ◊math{11} from its stream because ◊math{Y} will not need them anymore.
+    ◊math{X} also includes signals for the newly received messages into the ◊math{X → Y} stream header and removes an obsolete signal for message ◊math{Y◊sub{1}} it consumed before.
+  }
+  ◊p{◊(embed-svg "images/08-signals.svg")}
+}
+◊p{
+  Signals and stream messages are like snakes eating each other tails on the ◊a[#:href "https://theneverendingstory.fandom.com/wiki/Auryn"]{Auryn}: the sender drops messages when it sees signals, and the receiver drops signals when it sees stream bounds.
```
Comment (alin-at-dfinity) [inline posts/08-ic-xnet.html.pm:164]:
```suggestion
  Signals and stream messages are like snakes eating each other's tails on the ◊a[#:href "https://theneverendingstory.fandom.com/wiki/Auryn"]{Auryn}: the sender drops messages when it sees signals, and the receiver drops signals when it sees stream bounds advancing.
```

File: posts/08-ic-xnet.html.pm [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    It randomly contacts nodes from other subnets , assigning higher weights to nodes with lower latency.
+  } peers to reduce the message delivery latency.
+}
+}
+
+◊section{
+◊section-title["block-payloads"]{Block payloads}
+◊p{
+  The job of the consensus protocol is to aggregate messages from the outside world and pack them into a neat block.
+  The consensus includes several types of messages into blocks, such as user ingress messages, Bitcoin transactions (for subnets with enabled Bitcoin integration), and inter-canister messages from other subnets.
```
[the remark this answers is not in the document]
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:56]: Done.

File: posts/08-ic-xnet.html.pm [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+  We call messages paired with the data required for their validation a ◊em{payload}.
+  We call components that pull payloads from the network ◊em{payload builders}.
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-payloads"]{The consensus algorithm aggregates messages from the outside world into blocks.}
+  ◊p{◊(embed-svg "images/08-block-payloads.svg")}
+}
+◊p{
+  XNet payload builder pulls messages from nodes assigned to other subnets using a simple HTTP protocol.
+  ◊em{XNet endpoint} is a component that serves messages destined for other subnets over secure TLS connections, accepting connections only from other nodes◊sidenote["sn-xnet-tls"]{This measure does not imply privacy because malicious nodes can access the data; but ensures that network providers cannot read the messages.}.
```
[the remark this answers is not in the document]
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:66]: The original placement of "only" is more correct according to my book. Not that "only" modifies the thing that directly follows.
https://www.wikihow.com/Use-the-Word-%22Only%22-Correctly

My semantics: `(accepting connections (only (from other nodes))`
Your semantics: `((only accepting) connections (from other nodes))`

File: posts/08-ic-xnet.html.pm (hunk 1 above)
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:84]: Good point. I replaced "conflates" with "interleaves".

File: posts/08-ic-xnet.html.pm [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+  Signals are a part of the XNet payload specifying the prefix of the ◊em{reverse} stream that the sending subnet can drop.
+  When a node from subnet ◊math{X} fetches XNet payload from a node from subnet ◊math{Y}, in addition to actual messages, the ◊math{Y} node includes the ◊em{stream header}.
+  The stream header describes the state of the ◊math{X ↔ Y} communication:
+}
+◊ul[#:class "arrows"]{
+  ◊li{
+    The full range of message indices in the forward stream ◊math{X → Y}.
+  }
+  ◊li{
+    The signals for the ◊em{reverse} stream (◊math{Y → X}): for each message index in the reverse stream, ◊math{Y} tells whether ◊math{X} can garbage collect the message (an ◊code{ACK} signal) or should reschedule the message (a ◊code{REJECT} signal).
```
[the remark this answers is not in the document]
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:138]: Done.

File: posts/08-ic-xnet.html.pm [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    A ◊code{REJECT} signal indicates that the destination canister moved, so ◊math{X} should insert the message into another stream.
```
[the remark this answers is not in the document]
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:139]: Done.

File: posts/08-ic-xnet.html.pm (hunk 3 above)
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:164]: Done.

File: posts/08-ic-xnet.html.pm (hunk 2 above)
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:144]: Done.

File: posts/08-ic-xnet.html.pm [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    The data flow in the XNet protocol.
+    Subnet ◊math{Y} produces a signed stream of messages for subnet X and exposes this stream via an HTTP endpoint.
+    Subnet ◊math{X} pulls messages from one of the nodes on subnet Y.
+  }
+  ◊p{◊(embed-svg "images/08-xnet-endpoint.svg")}
+}
+}
+
+◊section{
+◊section-title["message-streams"]{Message streams}
```
[the remark this answers is not in the document]
Comment (roman-kashitsyn) [reply posts/08-ic-xnet.html.pm:81]: Makes a lot of sense, done!

**Pull request merged**